render.sh

```diff
@@ -1,21 +1,61 @@
 #!/bin/bash
 
-URL=${1/https:\/\/www/http:\/\/xml}
-
-if hash curl 2>/dev/null ; then
-  curl -s $URL | xsltproc zeitoffline.xslt - > artikel.html
-elif hash wget 2>/dev/null ; then
-  wget $URL --quiet -O - | xsltproc zeitoffline.xslt - > artikel.html
-else
-  echo "Neither curl nor wget found. Aborting."
-  exit 1
-fi
-
-if hash xdg-open 2>/dev/null ; then
-  xdg-open artikel.html
-elif hash open 2>/dev/null ; then
-  open artikel.html
-else
-  echo "Neither xdb-open nor open found. Aborting, as I don't know how to open your browser with artikel.html"
-  exit 1
-fi
+has_program() {
+    hash $1 2> /dev/null
+    return $?
+}
+
+get() {
+    local url=${1/https:\/\/www/http:\/\/xml}
+
+    if has_program curl; then
+        curl -s $url
+    elif has_program wget; then
+        wget $url --quiet -O -
+    else
+        echo "Neither curl nor wget found. Aborting."
+        usage
+        exit 1
+    fi
+}
+
+convert() {
+    xsltproc zeitoffline.xslt -
+}
+
+view() {
+    local document=$1
+    if has_program xxdg-open; then
+        xdg-open $document
+    elif has_program open; then
+        open $document
+    else
+        echo "Neither xdg-open nor open found. Aborting, as I don't know how to open your browser"
+        usage
+        exit 1
+    fi
+}
+
+usage() {
+    cat <<EOF
+Usage: $0 URL
+
+Converts articles from zeit.de from XML to simple HTML.
+
+Requires either curl or wget for downloading, xslt for conversion and
+xdg-open or open for viewing.
+EOF
+}
+
+main() {
+    local filename=$(mktemp -t zeitoffline.XXXXXX.html)
```

Comment:
I'm afraid this doesn't work with BSD style mktemp.

Reply:
Also, when should we clean up the temporary files?

Reply:
Yes, on OS X I get: No application knows how to open /var/folders/pp/80_6xcgd5qzdblvr9l1z_9v80000gn/T/zeitoffline.XXXXXX.html.iLSkCOd8.

Reply:
I did a quick research, but didn't find a satisfactory cross-platform way to create a temporary file with a suffix. We would have to resort to creating a temporary directory instead and place the file there. Cleaning that up will be even more annoying. Still, it's kind of unsatisfactory to have a global filename and therefore only view one URL at a time. Do you have any other ideas?

Reply:
On BSD like systems the generated temporary file could be renamed to $tmpfilename.html. In order to avoid handling GNU and BSD we could use something like "$(mktemp).html"

Reply:
That would sort of work, but goes against the grain of what `mktemp` does. The point of `mktemp` is to create files in `/tmp` securely, even in the presence of a malicious user on the same system, who will also have write access to `/tmp`. That's why `mktemp` does not only print a random filename, but also creates an empty file, so we get twice the garbage.

Reply:
Would renaming the file after creating avoid breaking this logic?